java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/client/ArrowFlightSqlClientHandler.java

```diff
@@ -92,11 +88,18 @@ private CallOption[] getOptions() {
    * @param flightInfo The {@link FlightInfo} instance from which to fetch results.
    * @return a {@code FlightStream} of results.
    */
-  public List<FlightStream> getStreams(final FlightInfo flightInfo) {
-    return flightInfo.getEndpoints().stream()
-        .map(FlightEndpoint::getTicket)
-        .map(ticket -> sqlClient.getStream(ticket, getOptions()))
-        .collect(Collectors.toList());
+  public List<FlightStream> getStreams(final FlightInfo flightInfo) throws RuntimeException {
+    try {
+      ArrayList<FlightStream> streams = new ArrayList<>();
+      for (FlightEndpoint ep : flightInfo.getEndpoints()) {
+        URI uri = ep.getLocations().size() > 0 ? ep.getLocations().get(0).getUri() : null;
+        FlightSqlClient sqlClient = this.connectionManager.getConnection(uri);
+        streams.add(sqlClient.getStream(ep.getTicket(), getOptions()));
```

Comment:
It's not really apparent how the sqlClient gets returned to the pool. It should get returned after the rest of the driver is done with the stream.

Reply:
Without trying to excuse he behavior, I was following the existing pattern: the builder created a `FlightSqlClient` and passed ownership of it to `ArrowFlightSqlClientHandler`, which closes it when it is closed itself.

This PR changes it so that: the builder creates a `ConnectionManager` and passes ownership of it to `ArrowFlightSqlClientHandler` which closes it (and it closes all connections) when it is closed itself.

I see that the resulting streams are `AutoClosable`. I think we could wrap them in an AOP proxy and return the client to the pool when the stream is closed. To be honest, calling what is there a "pool" is pretty strong. Presently there's no notion of outstanding leases, return, etc. If this behavior is desired, I can look into an `apache-commons` `Pool` or similar.

Reply:
I started to make a `FlightStreamWrapper` that could track the number of open `FlightStreams` for a given client, but I'm at a bit of a loss as to the intended life-cycle:

1. Do we want to open `FlightClient`s to each of these endpoints and keep them open?
2. Should do we want to close the `FlightClient` to an endpoint as soon as it's stream is closed?
3. Do we want to put in some kind of timeout where we close it after some amount of inactivity?

Just let me know what's desired and I can make it happen.

Reply:
Sorry for the late response. I've been thinking about this and I think the right thing to do is to actually change the protocol such that when a Flight Server responds to getFlightInfo, it can optionally add an optimization hint telling the caller that endpoints are long-lived and potentially re-usable. Perhaps indicating as well if contacting an endpoint can just re-use information from the originating request (such as cookies or authentication tokens).

For this PR, let's pessimistically assume that we should not reuse FlightClients and close them with the stream.  And discuss how/if the server should provide lifetime information about endpoints in the dev mailing list.

Reply:
I'll update the PR shortly. I think the easiest way to do this is:

1. Update `ConnectionManager.acquireConnection()` to create a new `FlightSqlClient` each time and return ownership
2. Create a `FlightStreamWrapper` class that intercepts the `close()` method
3. Update `ArrowFlightSqlClientHandler.getStreams()` to wrap the returned streams in `FlightStreamWrapper`s
4. Have `FlightStreamWrapper` call `close` on the `FlightSqlClient` when `close()` is called on the stream

Sound good?

Reply:
Yeah this sounds like a good plan.